zkEVM - add json-rpc doc (#585)
0xPolygon/polygon-docs: docs/zkEVM/architecture/proving-system/json-rpc-to-proof.md

This document describes the interaction of users with the Polygon zkEVM network through a component known as zkEVM JSON-RPC.

It provides,

- An overview of how L2 blocks are generated.
- A detailed explanation of how transactions sent through JSON-RPC are processed, including their validation before being stored in the PoolDB.

The main components involved in this process are the JSON-RPC, PoolDB, sequencer, executor, prover, HashDB, and Node StateDB.

## zkEVM JSON-RPC

The Ethereum JSON-RPC is a standardized collection of methods used by all Ethereum nodes, acting as the main interface for users to interact with the network.

The zkEVM JSON-RPC is an interface compatible with the Ethereum JSON-RPC for any layer 2 system.

> It replicates the same endpoints as the Ethereum JSON-RPC, and similarly receives queries and provides answers to users. 

Reviewer: ```suggestion
It implements the same endpoints as the Ethereum JSON-RPC, and similarly receives queries and provides answers to users. 
```

The user’s retrieved data comprises all information related to the block, including its transactions.

Although the zkEVM JSON-RPC incorporates all Ethereum JSON-RPC methods, some of its answers to queries are only relevant to the zkEVM context.

The user can request for the latest block as shown below:

![Figure: How user queries](../../../img/zkEVM/rpc-how-user-makes-query.png)

The zkEVM JSON-RPC not only includes methods from the Ethereum JSON-RPC, but it also offers additional methods to handle zkEVM-specific functionalities. 

The complete list of available endpoints and their current implementation status can be obtained in the [json-rpc-endpoints.md](https://github.com/0xPolygonHermez/zkevm-node/blob/develop/docs/json-rpc-endpoints.md).

The API specification for the zkEVM endpoints, which follows the OpenRPC standard, is available in the [endpoints_zkevm.openrpc.json](https://github.com/0xPolygonHermez/zkevm-node/blob/develop/jsonrpc/endpoints_zkevm.openrpc.json) file.

Ethereum JSON-RPC responds with the most recently created L1 block when requesting the last block. 

zkEVM batches transition through various states: trusted, virtual, and consolidated.

> We explore next the expected form of data in the zkEVM JSON-RPC's responses to queries for the latest L2 block.

Reviewer: ```suggestion
Next, we explore the expected format of data in the zkEVM JSON-RPC's responses to queries for the latest L2 block.
```

Consequently, every batch contains as many blocks as transactions.

Also, batch data includes only one timestamp which is shared across all the blocks within the batch.

The figure below illustrates the batch structure.

![Figure: ](../../../img/zkEVM/rpc-batch-w-several-blocks.png)

### Etrog upgrade (ForkID 6)

The implementation approach of the Dragonfruit upgrade and prior zkEVM versions assigns one ordered transaction to a block, reducing delay to a minimum threshold of the $\mathtt{order\_a\_transaction\_time}$.

The Dragonfruit upgrade approach has some drawbacks:

- Bloated databases: Since there are as many blocks as there are transactions in each batch, it results in a significant amount of data in the database.
- Incompatibility with dApps: It cannot offer a method for assigning a unique timestamp to each block within a batch, causing breaks in Dapps that rely on this parameter to determine the timing of actions performed by smart contracts.

The Etrog upgrade (which is ForkID 6) addresses these two issues.

That is, each L2 block consists of one or more transactions, and each block has its own unique timestamp.

> With this implementation approach, the sequencer needs to make ensure that block construction has a shorter timeout than the batch generation.

Reviewer: ```suggestion
With this implementation approach, the sequencer needs to ensure that block construction has a shorter timeout than the batch generation.
```

The sequencer can modify the timestamp of the various blocks in a batch by utilizing a specific transaction as a marker, dubbed `changeL2Block`. 

This transaction is included in the batch to indicate a change from one block to the next 

> in blocks and is responsible for modifying the timestamp and the L2 block number. 

Reviewer: Unexpected gap here. Please check and correct.

The figure below depicts the structure of the Etrog upgrade's batches.

![Figure: Fork-ID 6 - batch approach](../../../img/zkEVM/rpc-batch-approach-forkid6.png)

Read the [Etrog upgrade document](../protocol/etrog-upgrade.md) for further details and how it differs from the Dragonfruit upgrade.

## Custom zkEVM endpoints

When a user requests information from the JSON-RPC regarding the latest block, the response shows the most recent L2 block.

The L2 block contains the most recent transactions approved by the trusted sequencer.

The information is retrieved with the same RPC call as in Ethereum.

> Users are not really concerned with the transaction being in a block, but rather when the batch transitions to a new state.

Reviewer: ```suggestion
Users are less concerned with the transaction being in a block, and more with when the newly generated batch results in a state transition.
```

The zkEVM protocol has additional endpoints for retrieving various information pertaining to the status of the L2 block.

> For example, the user may query whether a block is virtualized or not by using a specific endpoint as shown in the figure below.

Reviewer: Please ensure the definition of the 'virtualized' status for a block is available to the reader.

![Figure: Checking if block is virtualized](../../../img/zkEVM/rpc-check-if-block-virtual.png)

Here is a list of custom zkEVM endpoints, each accompanied by a brief description:

- $\mathtt{zkevm\_consolidatedBlockNumber}$: Returns the latest block number within the last verified batch.

> - $\mathtt{zkevm\_isBlockVirtualized}$: Returns true if the provided block number is in a virtualized batch.

Reviewer: ```suggestion
- $\mathtt{zkevm\_isBlockVirtualized}$: Returns `true` if the provided block number is in a virtualized batch.
```

> - $\mathtt{zkevm\_isBlockConsolidated}$: Returns true if the provided block number is in a consolidated batch.

Reviewer: ```suggestion
- $\mathtt{zkevm\_isBlockConsolidated}$: Returns `true` if the provided block number is in a consolidated batch.
```

- $\mathtt{zkevm\_batchNumber}$: Returns the latest batch number.

- $\mathtt{zkevm\_virtualBatchNumber}$: Returns the batch number of the latest virtual batch.

- $\mathtt{zkevm\_verifiedBatchNumber}$: Returns the batch number of the latest verified batch.

- $\mathtt{zkevm\_batchNumberByBlockNumber}$: Returns the batch number of the batch containing the given block.

> - $\mathtt{zkevm\_getBatchByNumber}$: Gets the info of a batch given its number.

Reviewer: ```suggestion
- $\mathtt{zkevm\_getBatchByNumber}$: Fetches the available info for a batch based on the specified batch number.
```

It therefore conducts initial validations on transactions.

If any of these validations fail, an error is sent to the JSON-RPC component, which then forwards it to the user.

For valid transactions, the pool conducts pre-execution in which it estimates transaction costs by using the current state root, which may be different when the transaction is ultimately ordered.

As illustrated in the figure below, validation is the first step that takes place in the pool, invoking the `validateTx()` function.

The figure below depicts the submission process.

![Figure: Sending txs to the system](../../../img/zkEVM/rpc-sending-tx-to-system.png)

In the `validateTx()` function performs the following preliminary checks:

1. The transaction IP address has a valid format.

2. The transaction fields are properly signed (in both current and pre-EIP-155). EIP- 155 states that we must include the `chainID` in the hash of the data to be signed (which is an anti-replay attack protection).

3. The transaction’s `chainID` is the same as the pool’s `chainID` (which is the `chainID` of the L2 Network) whenever `chainID` is not zero.

> 4. The transaction string has an encoding that is accepted by the zkEVM. (See the next subsection for more details on this encoding.)

Reviewer: Adding an internal link here to jump to the next sub-section might be helpful.

7. The transaction’s value is not negative (which can be the case since we are passing parameters over an API).

8. The transaction sender’s address is not blacklisted. That is, it is not blocked by the zkEVM network.

9. The transaction preserves the nonce ordering of the account.

10. The transaction sender account has not exceeded the allowed maximum number of transactions per user to be contained in the pool.

11. The pool is not full (currently, the maximum number of elements in the queue of pending transactions is 1024).

12. The gas price is not lower than the set _minimum gas price_. This is explained in more detail in the [Effective gas price section](../effective-gas/index.md).

13. The transaction sender account has enough funds to cover the costs
    
    $$
    \mathtt{value} + \mathtt{GasPrice} \cdot \mathtt{GasLimit}
    $$

14. The computed _intrinsic gas_ is greater than the provided gas. The _intrinsic gas_ of a transaction measures the required gas in terms of the amount of transactional data plus the starting gas for the raw transaction which is currently of $21000$, or $53000$ in the case of a contract creation transaction.

> 15. The current transaction gasprice is higher than the other transactions in the PoolDB with the same nonce and from. This is because a transaction cannot be replaced with another with less gasprice.

Reviewer: ```suggestion
15. The current transaction `GasPrice` is higher than the other transactions in the PoolDB with the same nonce and `from` address. This is because a transaction cannot be replaced with another with lower `GasPrice`.
```
